runtime/protocol/http-client-engines/http-client-engine-crt/common/src/aws/smithy/kotlin/runtime/http/engine/crt/RequestUtil.kt

```diff
@@ -109,3 +113,24 @@ internal suspend fun HttpStream.sendChunkedBody(body: HttpBody) {
         else -> error("sendChunkedBody should not be called for non-chunked body types")
     }
 }
+
+internal inline fun <T> mapCrtException(block: () -> T): T =
+    try {
+        block()
+    } catch (ex: CrtRuntimeException) {
+        throw HttpException(ex.message ?: ex.errorDescription, ex, errorCode = mapCrtErrorCode(ex.errorCode))
+    }
+
+// do this by name rather than error code as it's difficult to map error codes on JVM side and would be prone to breaking
+// if new errors are added to the various aws-c-* lib enum blocks.
+//
+// See:
+// IO Errors: https://github.com/awslabs/aws-c-io/blob/v0.13.19/include/aws/io/io.h#L89
+// HTTP Errors: https://github.com/awslabs/aws-c-http/blob/v0.7.6/include/aws/http/http.h#L15
+internal fun mapCrtErrorCode(code: Int): HttpErrorCode = when (CRT.errorName(code)) {
+    "AWS_IO_SOCKET_TIMEOUT" -> HttpErrorCode.SOCKET_TIMEOUT
+    "AWS_IO_TLS_ERROR_NEGOTIATION_FAILURE", "AWS_IO_TLS_ERROR_NOT_NEGOTIATED" -> HttpErrorCode.TLS_NEGOTIATION_ERROR
+    "AWS_ERROR_HTTP_UNSUPPORTED_PROTOCOL" -> HttpErrorCode.PROTOCOL_NEGOTIATION_ERROR
+    "AWS_ERROR_HTTP_CONNECTION_CLOSED" -> HttpErrorCode.CONNECTION_CLOSED
+    else -> HttpErrorCode.SDK_UNKNOWN
```

Comment:
Consider adding:

IO Exceptions:
```kotlin
AWS_IO_BROKEN_PIPE -> HttpErrorCode.CONNECTION_CLOSED
AWS_IO_TLS_NEGOTIATION_TIMEOUT -> HttpErrorCode.TLS_NEGOTIATION_TIMEOUT
AWS_IO_SOCKET_NOT_CONNECTED -> HttpErrorCode.CONNECT_TIMEOUT
AWS_IO_SOCKET_CLOSED -> HttpErrorCode.CONNECTION_CLOSED
AWS_IO_TLS_DIGEST_ALGORITHM_UNSUPPORTED -> HttpErrorCode.TLS_NEGOTIATION_ERROR
AWS_IO_TLS_SIGNATURE_ALGORITHM_UNSUPPORTED -> HttpErrorCode.TLS_NEGOTIATION_ERROR
```
 
HTTP Errors:
```kotlin
AWS_ERROR_HTTP_SERVER_CLOSED -> HttpErrorCode.CONNECTION_CLOSED
AWS_ERROR_HTTP_STREAM_HAS_COMPLETED -> HttpErrorCode.CONNECTION_CLOSED
```

Reply:
Good callout, I wasn't sure on some of these and was admittedly overly conservative but I think most of them look fine. In particular, `AWS_IO_BROKEN_PIPE`, `AWS_IO_SOCKET_NOT_CONNECTED`, `AWS_IO_SOCKET_CLOSED`, and `AWS_ERROR_HTTP_STREAM_HAS_COMPLETED` are the ones I am uncertain they map cleanly. I think we could take them for now though as suggested unless anyone else has objections.